Empties subject title of DMs and GDMs when converting from Slack to Zulip

elski569 opened the pull request:
The pull requests consists of three parts/commits:
- Two assertions that checks if topic_name is empty string for private messages.
- A condition in `slack.py` that check if message is private and sets `topic_name=""`.
- A migration that changes previously converted messages and sets subject="" for DMs and GDMs.

Fixes: [#29466](https://github.com/zulip/zulip/issues/29466 )

timabbott commented on zerver/data_import/slack.py:
```diff
@@ -999,6 +999,10 @@ def channel_message_to_zerver_message(
 
+        # The topic name is empty string for DMs and GDMs.
+        if is_private:
+            topic_name = ""
```
Is the scenario where this might happen a Slack thread in a DM?

elski569 replied:
> Is the scenario where this might happen a Slack thread in a DM?

I would say that the scenario where this might happen is either when the message is a DM, GDM or a thread in a DM. I understand it as if a thread in a DM should be "flattened" into the main conversation in Zulip. Did I understand your question correctly? Do you think I have the wrong approach?

PieterCK replied:
Yes, this would cover DM, group DM, and thread in group DM (this will have an additional `thread_ts` field)

timabbott replied:
Could this logic be moved to `build_message`, so that it's done correctly for all data import tools, including future ones we might write?

elski569 replied:
I could potentially add something like this to build_message: 
`  recipient = Recipient.objects.get(id=recipient_id)
    # Now you can access the type_id from the recipient object
    recipient_type_id = recipient.type_id
    if recipient_type_id == 1 or recipient_type_id == 3:
        topic_name = ""`
what do you think?

timabbott replied:
I'm a little confused because we shouldn't be doing any writes to the database from these tools, so `Recipient.objects.get(id=recipient_id)` doesn't make any sense. 

But I think `build_message` could reasonably clear the `topic` field if it's not the `""` on DMs, right? The caller should be passing something that clearly indicates that, or could easily do so if it isn't.

elski569 replied:
Yes, we could likely clear the `topic_name` in `build_message`. However, the reason for my solution is that I couldn’t find anything in build_message that indicates whether the message is a DM. Perhaps we could add a field like `is_private` to make it possible to clear the topic in `build_message`. What do you think?

timabbott replied:
Adding a parameter seems fine, yes.

timabbott commented on zerver/migrations/0624_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,34 @@
+    ) | Message.objects.filter(
+        subject__regex=slack_thread_pattern,
+        recipient__type__in=[RECIPIENT_PERSONAL, RECIPIENT_HUDDLE],
+    )
```
Is it possible to filter on `sending_client` to help the server not have to process every message to run this migration?

elski569 replied:
> Is it possible to filter on `sending_client` to help the server not have to process every message to run this migration?

I understand it as imported messages have `sending_client = 1` and if that is correct I can add it to the filter. What do you think? Would that be a good solution?

PieterCK replied:
Since we're dealing with `Message` table I wonder if we need to process this in batches like `0257_fix_has_link_attribute`? though if the `sending_client` field for imported messages is `Internal` that'd narrow the size by quite a bit.

timabbott replied:
Yes, we should probably process in batches. I think `1` is not correct, since `Client` objects (and everything else in the imports) get renumbered during the import process. You'll need to look it up via the `populate_db` client name, instead.

elski569 commented:
@timabbott Do you think this pull request contains everything it should based on the description of the issue?

elski569 commented:
@timabbott 
Could you possibly look through this PR any time soon? I would appreciate any feedback or suggestions if I have the wrong approach. Thank you in advance :)

timabbott commented:
@PieterCK do you have time to review this one? It seems very much up your alley.

PieterCK commented:
Ok, I think I can review this tomorrow.

PieterCK reviewed (commented):
Great, thanks for working on this fix! A couple of comments on commit organization:
- The first commit can be squashed down to the second commit,
  From the [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages):
  >It should pass tests (so test updates needed by a change should be in the same commit as the original change, not a separate “fix the tests that were broken by the last commit” commit).
- The last commit in the series should have `Fixes #123` in its description.
- Start the description with an imperative verb, e.g., “fix”, “add”, “change”, “rename”, etc.

PieterCK commented:
@elski569 Let me know if any points are unclear or if you have any questions!

elski569 commented:
Thank you for the review, @PieterCK! I’ve incorporated your proposed changes into the PR.

PieterCK commented:
Thanks for making these changes!

Ok, so I have done some manual testing. Probably imported messages have `populate_db` and not `Internal` as their `sending_client`, shell:
```
In [22]: Message.objects.filter(sending_client=60, realm=22).count()
Out[22]: 1

In [23]: Client.objects.get(name="populate_db").id
Out[23]: 63

In [24]: Message.objects.filter(sending_client=63, realm=22)
Out[24]: <QuerySet [<Message: general / imported from Slack / <UserProfile: pieterceka123@gmail.com <Realm: zazaasdasddda 22>>>, <Message: general / imported from Slack / <UserProfile: pieterceka123@gmail.com <Realm: zazaasdasddda 22>>>, <Message: general / imported from Slack / <UserProfile: pieterceka123@gmail.com <Realm: zazaasdasddda 22>>>, <Message: general / imported from Slack / <UserProfile: pieterceka123@gmail.com <Realm: zazaasdasddda 22>>>, <Message: general / imported from Slack / <UserProfile: pieterceka123@gmail.com <Realm: zazaasdasddda 22>>>]>
```
  
  I also added a couple of manual messages after importing, and those indeed have `website` as their `sending_client`.

PieterCK commented:
I think @timabbott can take over for the next round of review at this point. A couple of things I wanted to highlight:  
- I didn’t manually import a Slack export with all the relevant cases (DMs, GDMs, etc.) to test the migration logic.  
- I did manually import a Slack export to check the imported messages’ `sending_value`. The result can be found here: https://github.com/zulip/zulip/pull/32298#issuecomment-2538857640  

Once the `sending_value` in the migration cleanup logic has been updated, the PR looks good to me. That said, I’m not very familiar with all the considerations we need to account for in the database cleanup logic here, so Tim might want to give it extra attention.

PieterCK commented:
@zulipbot add "integration review"

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+    Message = apps.get_model("zerver", "Message")
+
+    # Fetch the ID of "populate_db" client which is the sending_client of imported messages
+    internal_client, _ = Client.objects.get_or_create(name="populate_db")
```
Oh dear, we do seem to have `zulip_message_dict["sending_client"] = 1`. That is a very odd choice, but I guess it's a preexisting issue.

timabbott replied:
Started https://chat.zulip.org/#narrow/channel/3-backend/topic/sending_client.20for.20data.20import/near/2004133 to discuss that, but I believe your logic here is accurate, however horrifying it is that it is accurate :).

elski569 replied:
Yes,` zulip_message_dict["sending_client"] = 1 ` in `import_util.py` is why I initially had `sending_client_id =1` in my migration. However, now the logic seems to be correct so I'll let it be.

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+        id__lte=id_end,
+        sending_client_id=internal_client.id,
+        recipient__type__in=[RECIPIENT_PERSONAL, RECIPIENT_HUDDLE],
+    ).filter(Q(subject="Imported from Slack") | Q(subject__regex=slack_thread_pattern))
```
Why not just filter on `.exclude(subject="")` here? That is probably faster and avoids making assumptions about the naming patterns not having changed.

elski569 replied:
That definitely seems like a better choice. Thank you :)

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+from django.db.migrations.state import StateApps
+from django.db.models import Q
+
+BATCH_SIZE = 1000
```
You want 10000; most messages will not match this filter.

elski569 replied:
Great point which I didn't consider. I have increased the batch size in my current PR.

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+    messages_to_update.update(subject="")
+
+    if id_end % 1000 == 0:
+        print(f"Processed messages up to ID {id_end} / {last_id}")
```
I think this `if` statement is useless?

timabbott replied:
Also you can put this print in the caller, to avoid passing `last_id` into this function.

elski569 replied:
That’s great input! I deleted the statement and moved the print to follow the processing of each batch in the caller.

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+    while id_range_upper_bound <= last_id:
+        process_batch(apps, id_range_lower_bound, id_range_upper_bound, last_id)
+
+        id_range_lower_bound = id_range_upper_bound + 1
```
I think you want to skip the `+1` if you want these to not advance to values like 5005

elski569 replied:
Agree, I've considered this in my new PR.

timabbott commented on zerver/migrations/0641_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,70 @@
+
+    # Handle any remaining messages
+    if last_id > id_range_lower_bound:
+        process_batch(apps, id_range_lower_bound, last_id, last_id)
```
You can probably skip this if you just make your while loop compare `id_range_lower_bound` to `last_id` instead?

elski569 replied:
Yes, I've now made an attempt to include it in the while-loop instead.

timabbott commented:
@elski569 it looks like you pushed changes -- can you respond to each of the commands on how you addressed them, and post a summary of your understanding of the state of then PR? Check the contributing guide for more detailed documentation.

elski569 commented:
Currently, I believe the PR is done in terms of the migration. I've considered all your comments and increased `BATCH_SIZE`, optimized the while-loop to include the last batch of messages, removed unnecessary if-statement and moved the print to the caller in the while-loop. Now we have to decide which approach to take if we want to set `topic_name=""` in build_message instead. I propose we add the field `is_private` to `build_message` which would enable us to decide if the message is a DM or not.

timabbott commented on zerver/migrations/0642_empty_subject_for_slack_dms.py:
```diff
@@ -0,0 +1,62 @@
+    Message = apps.get_model("zerver", "Message")
+
+    # Fetch the ID of "populate_db" client which is the sending_client of imported messages
+    internal_client, _ = Client.objects.get_or_create(name="populate_db")
```
Let's call this `slack_import_client`

timabbott replied:
Also, don't do `get_or_create`; this migration should just return without doing anything if that client doesn't exist, just like the case for no messages existing. 

It also probably is best to fetch the client object in the caller and pass it into this helper.

elski569 replied:
Yes, I will change the name of client to `slack_import_client`, change to `Client.objects.get(name="populate_db")` and instead fetch client in caller.

timabbott commented:
@elski569 are you still working on this? If not, someone else can finish the project.

PieterCK commented:
I opened #33826 to follow up on this PR, it's ready for review

timabbott commented:
Closing as this was completed via #33826.
